# JSoC2015 Proposal


Efficient data structures and algorithms for sequence analysis in BioJulia
--------------------------------------------------------------------------

* Applicant: Kenta Sato (Email: bicycle1885@gmail.com, GitHub: @bicycle1885)
* Mentor: Daniel C. Jones (GitHub: @dcjones)


Summary
-------
Large-scale sequence analysis is a fundamental part of recent bioinformatics.
High-throughput DNA sequencers generate massive amounts of short DNA fragments from biological samples. These sequenced fragments often must be located in other long reference sequences for downstream analyses.
In order to find the matching locations for these short sequences in a long sequence like the human genome, efficient data structures and algorithms are indispensable.
In this project, under the BioJulia project (https://github.com/BioJulia), we will develop a full-text search and alignment library based on the FM-index[1].
The result will enhance the use of the Julia language for sequence analysis in the bioinformatics community and hence encourage more developers to write programs in the related fields of biology.


What I will make
----------------
I will implement a sequence aligner based on the FM-index to handle genome-scale sequence data from high-throughput sequencing.
The FM-index is the most commonly used technique to build fast and compact indices for full-text search, adopted by Burrows-Wheeler Aligner (BWA)[2] and bowtie2[3] when aligning massive numbers of short sequences to a large genome sequence.
The FM-index search is run over a text generated by applying the Burrows-Wheeler transform (BWT)[3] to the original text. The transformed text needs to be stored in a compact data structure that supports fast occurrence count of characters[5].
To fulfill the demands, I will develop Julia packages for two succinct data structures: bit vectors and Wavelet trees/matrices.
These data structures enable occurrence counts of characters for a sequence of arbitrary length in constant time.

Since the BioJulia project already has basic data structures to represent sequences, I will implement these tools on the top of them for seamless integration.
As for biological sequences, the alignment algorithm should allow small mismatches or gaps to be robust against noises and genetic variants.
I will consult other implementations to handle short mismatches well.


Demonstration
-------------
I have made two new packages required to create the search engine in order to show the implementability of this project:
* [IndexedBitVectors.jl](https://github.com/bicycle1885/IndexedBitVectors.jl)
* [WaveletMatrices.jl](https://github.com/bicycle1885/WaveletMatrices.jl)

IndexedBitVectors.jl implements a data structure like a bit vector, which allows some additional operations like bit count in constant time with small memory overhead.
WaveletMatrices.jl is built on the top of IndexedBitVectors.jl, supporting similar operations on 8bit elements rather boolean elements.

Using these data structures and the FM-index, I implemented a DNA fragment searcher in the human genome.
It can calculate the number of occurrences of the query fragment in a chromosome (about 51M length) for 10-50μ seconds while a linear searcher requires nearly x5000 time (50-100m seconds) for a short sequence.
Some examples are listed below:

*Search time in μs (best of three trials)*

| Query DNA Fragment | FM-index  | Linear Search |
|:------------------ | ---------:| -------------:|
| 'AAA'              |        13 |       128,992 |
| 'ACGTACGT'         |        27 |       111,608 |
| 'TCTGGAGCCAGAAGC'  |        45 |        82,140 |

For more details, please refer to the following links:
* https://github.com/bicycle1885/WaveletMatrices.jl#demo
* https://github.com/bicycle1885/WaveletMatrices.jl/blob/master/fmindex.jl


Plan
----
* June 15-30 - Choose and implement data structures.
    * There are several variants of succinct data structures. I will survey researches and try proposed data structures.
* July 1-14 - Implement efficient operations and FM-index.
    * Fast and low memory constructions are needed when handling large data. I will implement constructors, FM-index search, and other APIs at this point.
* July 15-31: Specialization for BioJulia.
    * Support search and alignment for biological sequences defined in BioJulia including short mismatches and gaps.
* September 1-15: Debug and performance tuning.
    * Brush up details of the program.


Impact
------
This will be a meaningful step toward sequence analysis because sequence search and alignment are indispensable building blocks of bioinformatics.
Almost all of the existing programs related to this kind of sequence analysis are implemented in C, C++, and Java, but simplicity and dynamic nature of Julia would be appealing if comparable performance can be achieved.
In addition, the succinct bit vector and the Wavelet matrix are applicable to implement various data structures including numeric sequences and graphs[6].
That means many developers outside the BioJulia project will be able to enjoy the benefit of efficient data structures.


About me
--------
I am a graduate school student at the University of Tokyo, Japan, studying bioinformatics and human genetics.
I have about 1.5 years’ experience of intensive Julia programming and have developed several packages in Julia:
* [DocOpt.jl](https://github.com/docopt/DocOpt.jl) - a command-line argument parser (registered in METADATA)
* [RandomForests.jl](https://github.com/bicycle1885/RandomForests.jl) - a machine learning method that builds lots of decision trees
* [Snappy.jl](https://github.com/bicycle1885/Snappy.jl) - a binding to a fast compression library
* [ANMS.jl](https://github.com/bicycle1885/ANMS.jl) - a derivative-free multivariate optimization method with adaptive parameters for high dimensional functions

I am one of founders of the [JuliaTokyo user group](http://julia.tokyo/) and we held meetups for three times in 2014 and 2015, the total turnout was about 150.


References
----------
1. Ferragina, Paolo, and Giovanni Manzini. "Opportunistic data structures with applications." Foundations of Computer Science, 2000. Proceedings. 41st Annual Symposium on. IEEE, 2000.
2. Li, Heng, and Richard Durbin. "Fast and accurate short read alignment with Burrows–Wheeler transform." Bioinformatics 25.14 (2009): 1754-1760.
3. Langmead, Ben, et al. "Ultrafast and memory-efficient alignment of short DNA sequences to the human genome." Genome Biol 10.3 (2009): R25.
4. Burrows, Michael, and David J. Wheeler. "A block-sorting lossless data compression algorithm." (1994).
5. Ferragina, Paolo, et al. "Compressed text indexes: From theory to practice."Journal of Experimental Algorithmics (JEA) 13 (2009): 12.
6. Claude, Francisco, and Gonzalo Navarro. "The wavelet matrix." String Processing and Information Retrieval. Springer Berlin Heidelberg, 2012.
